jinahub/indexers/searcher/compound/FaissPostgresSearcher/faisspsql.py

```diff
@@ -2,51 +2,122 @@
-        dump_path = dump_path or kwargs.get('runtime_args').get('dump_path')
         self.logger = get_logger(self)
+
+        self.total_shards = total_shards
+        if self.total_shards is None:
+            self.logger.warning(
+                'total_shards is None, rolling update '
+                'via PSQL import will not be possible.'
+            )
+
+        # when constructed from rolling update the dump_path is passed via a
+        # runtime_arg
+        dump_path = dump_path or kwargs.get('runtime_args').get('dump_path')
+        use_dump_func = use_dump_func or kwargs.get('runtime_args').get('use_dump_func')
+
         self._kv_indexer = None
         self._vec_indexer = None
-        if dump_path:
-            self._vec_indexer = FaissSearcher(dump_path=dump_path, **kwargs)
+        self._init_kwargs = kwargs
+
+        if dump_path is None and use_dump_func is None:
+            name = getattr(self.metas, 'name', self.__class__.__name__)
+            self.logger.warning(
+                f'No "dump_path" or "use_dump_func" provided '
+                f'for {name}. Use .rolling_update() to re-initialize...'
+            )
+
+        if use_dump_func:
             self._kv_indexer = PostgreSQLStorage(**kwargs)
+            dump_func = self._kv_indexer.get_snapshot
+            self._vec_indexer = FaissSearcher(dump_func=dump_func, **kwargs)
         else:
+            self._kv_indexer = PostgreSQLStorage(**kwargs)
+            self._vec_indexer = FaissSearcher(dump_path=dump_path, **kwargs)
+
+    @requests(on='/sync')
+    def sync(self, parameters: Dict, **kwargs):
+        """
+        Sync the data from the PSQLStorage into the FaissSearcher
+        """
+        use_delta = parameters.get('use_delta', False)
+        only_delta = parameters.get('only_delta', False)
+
+        if only_delta:
             self.logger.warning(
-                f'No dump path provided for {self}. Use .rolling_update() to re-initialize...'
+                'Syncing via delta method. This cannot guarantee consistency'
             )
+            pass
+            # we get the timestamp from time 0 and pass an argument to Indexer so
+            # that it can ignore the "delete"s
+
+        else:
+            self.logger.info('Syncing via snapshot...')
+            if self.total_shards:
+                dump_func = functools.partial(
+                    self._kv_indexer.get_snapshot, total_shards=self.total_shards
+                )
+                timestamp = self._kv_indexer.last_snapshot_timestamp
+                self._vec_indexer = FaissSearcher(
+                    dump_func=dump_func, **self._init_kwargs
+                )
+
+                if use_delta:
+                    self.logger.info(f'Now adding delta from timestamp {timestamp}')
+
+                    deltas = self._kv_indexer._get_delta(
+                        shard_id=0, total_shards=3, timestamp=timestamp
```

Comment:
Old code? I guess this needs to be some variable and not hard-coded `0` and `3`.

Reply:
Yes, fixed now